### PR TITLE
Feature/replace jsoncons

### PR DESCRIPTION
**Pull request by Pheubel**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/novelrt/NovelRT/blob/misc/templates/Contributing.md#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Replaces jsoncons with nlohmann json as json provider for NovelRT

**Is there an open issue that this resolves? If so, please [link it here](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).**
This PR partially resolves https://github.com/novelrt/NovelRT/issues/500, more specifically; jsoncons was causing issues when compiling for C++ 20. 

**What is the current behavior?** (You can also link to an open issue here)
The `DesktopResourceLoader` uses jsoncons for bson deserialization/serialization

**What is the new behavior (if this is a feature change)?**
The `DesktopResourceLoader` uses nlohmann json for bson deserialization/serialization

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
If users relied on jsoncons as json provider, they will have to either set up the inclusion themself, or convert their json code to work with nlohmann json.
If users relied on nlohmann json as json provider (indirectly provided through Fabulist) then they should not have to change their code.

**Other information**:
So far I only touched the CMake and the C++ code. I do not really know what exact alterations I would have to make to the [LICENCE-DIST.md](https://github.com/novelrt/NovelRT/blob/main/LICENCE-DIST.md) and the [build-system.yml](https://github.com/novelrt/NovelRT/blob/main/.github/workflows/build-system.yml).

### CODE REVIEW

**Comment by RubyNova** [comment]
The build system YML should not require changes but I'll leave that up to @capnkenny to confirm.

The licence distributable file requires you to add the licence for the new dependency, and remove the jsoncons licence assuming I added it originally, and that's it.

File: internal/CMakeLists.txt
Comment (capnkenny) [inline internal/CMakeLists.txt:3]: This should ideally be set inside of Fabulist's dependency CMakeLists, before the `MakeAvailable` call

**Review by capnkenny — COMMENTED** [review]
Overall seems good - just needs the licensing change, and I would move the one flag being set.

**Comment by capnkenny** [comment]
Just confirming - build system does not need to change and passes as-is

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (capnkenny) [inline src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:300]: Is there a reason this isn't a std::vector of the json object anymore?

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (capnkenny) [inline src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:307]: Is `member.type` already a uint32_t for nlohmann::json?

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (capnkenny) [inline src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:278]: What are the possibly `BinaryDataType`s here?

File: internal/nlohmannJson/CMakeLists.txt
Comment (capnkenny) [inline internal/nlohmannJson/CMakeLists.txt:34]: Was this taken from Fabulist's version of the nlohmann::json dependency CMakeLists?

**Review by capnkenny — COMMENTED** [review]
Just a few questions that need answering before approval or changes requested

File: internal/nlohmannJson/CMakeLists.txt
Comment (Pheubel) [reply internal/nlohmannJson/CMakeLists.txt:34]: yeah it was, i did bump the version, as fabulist is a little behind.

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (Pheubel) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:278]: If you mean how it's serialized: it's serialized as an integer by default,they can be serialized as strings as seen [here](https://json.nlohmann.me/features/enum_conversion/).

if you mean the values `BinaryDataType` can represent, then a full list can be seen [here](https://github.com/novelrt/NovelRT/blob/main/include/NovelRT/ResourceManagement/BinaryDataType.h). examples are:
* Boolean
* Int32
* Int64

File: internal/nlohmannJson/CMakeLists.txt
Comment (capnkenny) [reply internal/nlohmannJson/CMakeLists.txt:34]: Not a problem - my only ask is that we double check how this aligns with the Engine's compiler flags? I believe there's a difference between the two and I'm not sure which we prefer...

Cc @RubyNova for compiler flag input

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (capnkenny) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:278]: CC @RubyNova - should we be typing as UInt32 here? I see the static cast has been dropped, but I'm not sure if its necessary anymore (CI appears to pass but idk if we specifically test this)

File: internal/nlohmannJson/CMakeLists.txt
Comment (RubyNova) [reply internal/nlohmannJson/CMakeLists.txt:34]: Since it is an external dependency I am unsure if it matters what flags this dependency requires.

You're best off asking @FiniteReality .

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (Pheubel) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:300]: The json object behaves like std::vector, i feel like having it be a json object instead would make more sense in showing what it actually is.

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (Pheubel) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:307]: peeking at the [json header file](https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/conversions/to_json.hpp#L320), it does seem to already look at the underlying value in the enum, but i could be wrong. it is not the easiest library to look through

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (RubyNova) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:300]: To confirm, do we have a test covering this code path in the case the vector would have multiple entries?

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (RubyNova) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:278]: LGTM

File: internal/nlohmannJson/CMakeLists.txt
Comment (FiniteReality) [reply internal/nlohmannJson/CMakeLists.txt:34]: Ideally we'd be using this format going forward, if I understand correctly. The `/external:W0` here is just used to turn off warnings for external code.

**Review by RubyNova — APPROVED** [review]
All questions have been resolved, LGTM

File: src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp
Comment (FiniteReality) [reply src/NovelRT/ResourceManagement/Desktop/DesktopResourceLoader.cpp:307]: It's converted to an int by default, but it can be customised: https://json.nlohmann.me/features/enum_conversion/

I would suggest doing so, as this avoids potential reordering errors.

**Comment by FiniteReality** [comment]
Merging on @RubyNova's instructions. Fingers crossed this doesn't break everything :stuck_out_tongue_winking_eye:

**Pull request merged**
